File: src/networking/frame.zig
const std = @import("std");

pub const ZeonFrame = extern struct {
    status: ZeonFrameStatus,
    target_length: u64,
    fixed_buffer: [1024]u8,

    pub fn to_buffer(self: *@This(), length: ?u64) void {
        std.mem.writeInt(u8, self.fixed_buffer[0..1], @intFromEnum(self.status), std.builtin.Endian.Little);
        switch (self.status) {
            .Command, .Error, .Auth, .KeyExchange => {
                if (length == null) {
                    @panic("Length was expected but not null!");
                }
                std.mem.writeInt(u64, self.fixed_buffer[1..9], length.?, std.builtin.Endian.Little);
                self.target_length = length.?;
            },
            else => {
                @memset(self.fixed_buffer[1..1024], 0);
            }
        }
    }

    pub fn from_buffer(self: *@This()) void {
        self.status = @enumFromInt(std.mem.readInt(u8, self.fixed_buffer[0..1], std.builtin.Endian.Little));
        switch (self.status) {
            .Command, .Error, .Auth, .KeyExchange => {
                self.target_length = std.mem.readInt(u64, self.fixed_buffer[1..9], std.builtin.Endian.Little);
            },
            else => {}
        }
    }

    pub fn write_buffer(self: *@This(), buffer: *[1015]u8) void {
        @memcpy(self.fixed_buffer[9..], buffer);
    }

    pub fn read_buffer(self: *@This()) []u8 {
        return self.fixed_buffer[9..self.target_length + 9];
    }
};

pub const ZeonFrameStatus = enum(u8) {
    Ok = 0,
    Command = 1,
    Error = 2,
    Auth = 3,
    KeyExchange = 4,
};

File: _src/net/frame.cpp
#include <cstdint>
#include <cstring>
#include <bit>
#include <array>

#include <net/frame.hpp>

namespace ZeonDB::Net {

	void ZeonFrame::to_buffer(ZeonFrameStatus status, uint64_t length) {
		memcpy(this->buffer.data(), &status, sizeof(status));
		switch (status) {
			case ZeonFrameStatus::Command:
			case ZeonFrameStatus::Error:
			case ZeonFrameStatus::Auth:
				this->target_length = length;
				memcpy(this->buffer.data() + sizeof(status), &length, sizeof(uint64_t));
				break;
			default:
				memset(this->buffer.data() + 1, 0, 1023);
		}
	}

	void ZeonFrame::from_buffer() {
		memcpy(this->buffer.data(), &this->status, sizeof(ZeonFrameStatus));
		switch (this->status) {
			case ZeonFrameStatus::Command:
			case ZeonFrameStatus::Error:
			case ZeonFrameStatus::Auth:
				memcpy(this->buffer.data() + sizeof(ZeonFrameStatus), &this->target_length, sizeof(uint64_t));
				break;
			default:
				break;
		}
	}

	void ZeonFrame::write_buffer(const std::array<char, 1015>& buffer) {
		memcpy(this->buffer.data() + sizeof(ZeonFrameStatus) + sizeof(uint64_t), buffer.data(), 1015);
	}

	std::array<char, 1015> ZeonFrame::read_buffer() {
		std::array<char, 1015> arr;
		memcpy(arr.data(), this->buffer.data() + sizeof(ZeonFrameStatus) + sizeof(uint64_t), 1015);
		return arr;
	}

	char* ZeonFrame::get_buffer() {
		return this->buffer.data();
	}

	uint64_t ZeonFrame::get_length() {
		return this->target_length;
	}

	ZeonFrameStatus ZeonFrame::get_status() {
		return this->status;
	}
}
